core/arch/arm/mm/core_mmu_v7.c

```diff
@@ -641,53 +645,115 @@ bool core_mmu_user_mapping_is_active(void)
 	return read_ttbr0() != read_ttbr1();
 }
 
-static paddr_t map_page_memarea(struct tee_mmap_region *mm)
+static void print_mmap_area(const struct tee_mmap_region *mm __maybe_unused,
+				const char *str __maybe_unused)
+{
+	if (!(mm->attr & TEE_MATTR_VALID_BLOCK))
+		debug_print("%s [%08" PRIxVA " %08" PRIxVA "] not mapped",
+				str, mm->va, mm->va + mm->size);
+	else
+		debug_print("%s [%08" PRIxVA " %08" PRIxVA "] %s-%s-%s-%s",
+				str, mm->va, mm->va + mm->size,
+				mm->attr & (TEE_MATTR_CACHE_CACHED <<
+					TEE_MATTR_CACHE_SHIFT) ? "MEM" : "DEV",
+				mm->attr & TEE_MATTR_PW ? "RW" : "RO",
+				mm->attr & TEE_MATTR_PX ? "X" : "XN",
+				mm->attr & TEE_MATTR_SECURE ? "S" : "NS");
+}
+
+static paddr_t map_page_memarea(const struct tee_mmap_region *mm, uint32_t xlat)
 {
-	uint32_t *l2 = core_mmu_alloc_l2(mm->size);
+	uint32_t *l2;
 	size_t pg_idx;
 	uint32_t attr;
 
-	if (!l2)
-		panic("no l2 table");
+	if (!xlat)
+		l2 = core_mmu_alloc_l2(mm->size);
+	else
+		l2 = phys_to_virt(xlat & SECTION_PT_ATTR_MASK,
+				  MEM_AREA_TEE_RAM_RW_DATA);
```

Comment:
You mean something like an `assert(l2);` ?
By construction, core_mmu_alloc_l2() returns a valid address or panics.
By construction, if `xlat` is not null, it already stores a valid L2 table address (+ attributes/descc. type). `map_page_memarea_in_pgdirs()` will assert that attributes are valid. Maybe, in this case, we could assert here that the attribute do match ?